### ISSUE TITLE
Quarkus support

### ISSUE THREAD

**Issue by StephenOTT** [yahoo/elide#2279]
Support for Quarkus. 

Thoughts on what it would take?

**Comment by aklish** [comment]
Interesting idea.  

Elide includes a number of JAX-RS endpoints we could leverage with Quarkus:
 - [JSON-API](https://github.com/yahoo/elide/blob/master/elide-core/src/main/java/com/yahoo/elide/jsonapi/resources/JsonApiEndpoint.java)
 - [GraphQL](https://github.com/yahoo/elide/blob/master/elide-graphql/src/main/java/com/yahoo/elide/graphql/GraphQLEndpoint.java)
 - [Swagger](https://github.com/yahoo/elide/blob/master/elide-swagger/src/main/java/com/yahoo/elide/swagger/resources/DocEndpoint.java)

What remains is configuration.  Quarkus feels similar to spring boot in that configuration consists in creating beans for dependency injection.  Elide has the following configuration beans:

https://github.com/yahoo/elide/blob/master/elide-spring/elide-spring-boot-autoconfigure/src/main/java/com/yahoo/elide/spring/config/ElideAutoConfiguration.java

Spring creates some of the beans on its own like the EntityManagerFactory.  Those must be resolved another way with Quarkus.  Also, Elide does its own dependency injection of hooks, models, and checks.  It needs
an `Injector` to do that:

https://github.com/yahoo/elide/blob/master/elide-core/src/main/java/com/yahoo/elide/core/dictionary/Injector.java

A custom one of these will be needed that wraps the Quarkus CDI context.

Overall, I don't think it would be too hard to get something running on Quarkus that runs on the JVM.  There maybe more work to support GraalVM native images.

**Comment by aklish** [comment]
I see the description states _extension_.  Will need to think about what else might be required to make that work.

**Comment by StephenOTT** [comment]
Thanks for doing the initial review and providing your thoughts. Really appreciate the effort!!

My interest stemmed from working with quarkus and finding "missing" features in the graphql generation, jpa, etc. Elide seemed to be a great next step!

I would be happy to contribute and support the effort to make it happen. 

I have worked with micronaut as well: doing quarkus would likely not be a lost effort when/if doing micronaut as ~same concepts and stuff to deal with.

**Comment by aklish** [comment]
#2284 is all that is needed to unblock this I think.  We'll probably do another release next week.  

I think a minimum POC would be to:

- Use one of the quarkus template/examples to scaffold the deploy & runtime builds.
- Create a BuildStep in the deploy module that uses Jandex to scan for the annotations [here](https://github.com/yahoo/elide/pull/2284)
- Create a Recorder in the runtime module that creates CDI beans for a ClassScanner & EntityDictionary using the annotations found in the prior step.
- See if the dictionary is accessible in a running quarkus service

If that can all come together, I think the rest will be just fleshing out the other beans to get a complete service.

**Comment by StephenOTT** [comment]
Sounds great.  I will watch for the release and begin some testing!

**Comment by aklish** [comment]
@StephenOTT 

I started playing with this.  It is not turning out to be as much work as I thought it might be (thus far).

Here is the branch I'm working on: QuarkusScaffolding

It is pretty barebones, but I think the remainder needed is:

1. Hook up configuration so it controls Elide bean construction.
2. A lot of tests
3. Add and test other Elide features (async, aggregations, dynamic models, swagger, etc).  These are probably lower priority.

I'd like to get this to a place where CRUD for JSON-API and GraphQL are working reasonably well.  We don't use Quarkus in production here at Yahoo.  I would love to get your feedback using this if you have some cycles once it has reasonable test coverage. 

Also, we'll need to add a lot more Elide test models that cover use cases like:
- Security Checks
- Lifecycle Hooks
- Relationships
- Complex Attributes
- Inheritance
- GraphQL & JSON-API

Feel free to add models and tests to this branch that reflect your use cases if you have time/interest.

**Comment by StephenOTT** [comment]
For the tests: what are you worried about not working compared to non-quarkus build (with the existing tests)?

I have not done your level of analysis: so just thinking from my initial review: if the tests are checking if proper bean construction is in place/active, then what else do you want to check?

**Comment by aklish** [comment]
You have to register classes reflection.  I'm not sure what would happen if you didn't.  Maybe this only breaks native mode applications.  

I guess the bigger concern is running in native.  Lots of things might not work.

Maybe the solution here is to change Elide's integration test framework to run with Quarkus and that will cover all of the use cases.

**Comment by aklish** [comment]
Update:

I was able to get a draft extension for Elide and also build an example native application (boots in 20-50ms on my mac).  I've only tested very basic things including:
1. JSON-API post & get
2. GraphQL upsert & get
3. Swagger & Graphiql schemas

Every use case really needs tests in native mode because if there is any reflection in that code path (including in dependent libraries), the use case will not work.

You can check out the bare bones Elide extension on the branch QuarkusScaffolding

I've also created an example project here:
https://github.com/aklish/elide-quarkus-example

There is still a lot of work and a lot of bugs to be found

**Comment by aklish** [comment]
One comment.  To get the example project to work, the entire elide project needs to be built and installed locally (including the new elide-quarkus module).

You will also need to install the GraalVM, set the right environment variables, and install its native image (all documented elsewhere).

Java 11 is required for this branch.

If you try it out and have issues with native, let me know and I'll see if I can help.

**Comment by aklish** [comment]
```
bash-3.2$ ./target/elide-quarkus-example-1.0.0-SNAPSHOT-runner 
produceElide
Scanning classes
produceDictionary
Overriding data store
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2021-09-14 10:19:47,306 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.net.URL
2021-09-14 10:19:47,313 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.util.Date
2021-09-14 10:19:47,313 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.util.TimeZone
2021-09-14 10:19:47,313 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.sql.Time
2021-09-14 10:19:47,313 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.time.Instant
2021-09-14 10:19:47,313 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.sql.Timestamp
2021-09-14 10:19:47,313 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.time.OffsetDateTime
2021-09-14 10:19:47,313 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.sql.Date
2021-09-14 10:19:47,324 INFO  [io.quarkus] (main) elide-quarkus-example 1.0.0-SNAPSHOT native (powered by Quarkus 2.2.2.Final) started in 0.095s. Listening on: http://0.0.0.0:8080
2021-09-14 10:19:47,324 INFO  [io.quarkus] (main) Profile prod activated. 
2021-09-14 10:19:47,324 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, elide-extension, hibernate-orm, narayana-jta, resteasy, smallrye-context-propagation]
```

**Comment by aklish** [comment]
```
bash-3.2$ ./target/elide-quarkus-example-1.0.0-SNAPSHOT-runner 
produceElide
Scanning classes
produceDictionary
Overriding data store
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2021-09-14 10:20:45,989 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.net.URL
2021-09-14 10:20:46,015 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.util.Date
2021-09-14 10:20:46,015 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.util.TimeZone
2021-09-14 10:20:46,016 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.sql.Time
2021-09-14 10:20:46,016 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.time.Instant
2021-09-14 10:20:46,016 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.sql.Timestamp
2021-09-14 10:20:46,016 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.time.OffsetDateTime
2021-09-14 10:20:46,016 INFO  [com.yah.eli.Elide] (main) Registering serde for type : class java.sql.Date
2021-09-14 10:20:46,019 INFO  [io.quarkus] (main) elide-quarkus-example 1.0.0-SNAPSHOT native (powered by Quarkus 2.2.2.Final) started in 0.050s. Listening on: http://0.0.0.0:8080
2021-09-14 10:20:46,019 INFO  [io.quarkus] (main) Profile prod activated. 
2021-09-14 10:20:46,019 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, elide-extension, hibernate-orm, narayana-jta, resteasy, smallrye-context-propagation]
```

**Comment by aklish** [comment]
To build locally, you will need to run:
mvn clean install
mvn clean install -f elide-quarkus

**Comment by StephenOTT** [comment]
Thanks. Have not lost track of this. Should be back on this shortly!

**Comment by aklish** [comment]
No worries.  I'm taking a break from Quarkus to get GraphQL subscriptions out.  

What's really missing for Quarkus is a way to run native tests that work with CI/CD.  Then that list of tests can be expanded to make sure there aren't any hidden paths that leverage reflection and blow up in native mode.

**Comment by smsajid** [comment]
Any update on this? Is there any plan to publish this in the extensions registry of Quarkus?

**Comment by peteroyle** [comment]
Hi @smsajid, I can't speak for publishing to the registry, just a contributor here. But to put some sort of update out there - I'm still hopeful of helping to move this extension forward. Sadly the work stream I have that involves using Quarkus and Elide together has taken a temporary back seat. I hope to return to that as soon as possible. 

I agree that publishing to the registry might be useful, maybe the next logical area of development.
